File: cicd_dagger_contract/typescript/build.example.ts
/**
 * Build module for Goserv
 */

import { dag, Container, Directory, File, object, func } from "@dagger.io/dagger"

@object()
export class Build {
  /**
   * Build builds a multi-architecture Docker image and exports it as an OCI tarball
   *
   * @param source Source directory containing the project
   * @param releaseCandidate Whether to build as release candidate (appends -rc to version)
   * @returns File containing the build artifact
   */
  @func()
  async build(
    source: Directory,
    releaseCandidate: boolean = false
  ): Promise<File> {
    // Print message and return
    const output = await dag
      .container()
      .from("alpine:latest")
      .withExec(["echo", "This is the Build function"])
      .stdout()

    // Print to show the message
    console.log(output)

    // Return a dummy file since the function signature requires File
    return dag.container().from("alpine:latest").file("/etc/hostname")
  }
}

File: cicd_dagger_contract/python/build.example.py
"""Build module for Goserv"""

import dagger
from dagger import dag, function, object_type
from typing import Optional


@object_type
class Build:
    @function
    async def build(
        self,
        source: dagger.Directory,
        release_candidate: Optional[bool] = False,
    ) -> dagger.File:
        """Build builds a multi-architecture Docker image and exports it as an OCI tarball
        
        Args:
            source: Source directory containing the project
            release_candidate: Whether to build as release candidate (appends -rc to version)
        
        Returns:
            File containing the build artifact
        """
        # Print message and return
        output = await (
            dag.container()
            .from_("alpine:latest")
            .with_exec(["echo", "This is the Build function"])
            .stdout()
        )

        # Print to show the message
        print(output)

        # Return a dummy file since the function signature requires File
        return dag.container().from_("alpine:latest").file("/etc/hostname")
